src/txmempool.cpp

```diff
@@ -28,7 +28,7 @@ CTxMemPoolEntry::CTxMemPoolEntry(const CTransaction& _tx, const CAmount& _nFee,
     hadNoDependencies(poolHasNoInputsOf), inChainInputValue(_inChainInputValue),
     spendsCoinbase(_spendsCoinbase), sigOpCost(_sigOpsCost), lockPoints(lp)
 {
-    nTxCost = GetTransactionCost(_tx);
+    nTxCost = std::max(GetTransactionCost(_tx), (MAX_BLOCK_COST * sigOpCost + MAX_BLOCK_SIGOPS_COST - 1) / MAX_BLOCK_SIGOPS_COST);
```

Comment:
Rather than lose the caching of the cost, how about we just change GetTxSize() to return the max of these two quantities?  This way we have access to the actual cost so we can write smarter code in the future, while I believe GetTxSize() is already used everywhere (but we should confirm) for feerate calculations, included size/feerate of ancestors and descendants, both in the mempool and in the mining code.